packages/flutter_tools/lib/src/build_system/source.dart

```diff
@@ -0,0 +1,231 @@
+  List<int> bytesForVersion() {
+    if (_versionFile != null) {
+      return _versionFile.readAsBytesSync();
+    }
+    if (_fileSystemEntity is File) {
+      final File file = _fileSystemEntity;
+      return file.readAsBytesSync();
+    }
+    return _fileSystemEntity.statSync().modified.toIso8601String().codeUnits;
+  }
+}
+
+/// Collects sources for a [Target] into a single list of [FileSystemEntities].
+class SourceVisitor {
+  /// Create a new [SourceVisitor] from an [Environment].
+  SourceVisitor(this.environment, [this.inputs = true]);
+
+  /// The current environment.
+  final Environment environment;
+
+  /// Whether we are visiting inputs or outputs.
+  ///
+  /// Defaults to `true`.
+  final bool inputs;
+
+  /// The entities are populated after visiting each source.
+  final List<SourceFile> sources = <SourceFile>[];
+
+  /// Visit a [Source] which contains a function.
+  ///
+  /// The function is expected to produce a list of [FileSystemEntities]s.
+  void visitFunction(InputFunction function) {
+    sources.addAll(function(environment));
+  }
+
+  /// Visit a [Source] which contains a file uri.
+  ///
+  /// The uri may that may include constants defined in an [Environment].
+  void visitPattern(String pattern) {
+    // perform substitution of the environmental values and then
+    // of the local values.
+    final List<String> segments = <String>[];
+    final List<String> rawParts = pattern.split('/');
+    final bool hasWildcard = rawParts.last.contains('*');
+    final bool isDirectory = pattern.endsWith('/');
+    if (hasWildcard && isDirectory) {
+      throw Exception('wildcard patterns are not supported with directories.');
+    }
+    String wildcardFile;
+    if (hasWildcard) {
+      wildcardFile = rawParts.removeLast();
+    }
+    // If the pattern does not start with an env variable, then we have nothing
+    // to resolve it to, error out.
+    switch (rawParts.first) {
+      case Environment.kProjectDirectory:
+        segments.addAll(
+            fs.path.split(environment.projectDir.resolveSymbolicLinksSync()));
+        break;
+      case Environment.kBuildDirectory:
+        segments.addAll(fs.path.split(
+            environment.buildDir.resolveSymbolicLinksSync()));
+        break;
+      case Environment.kCacheDirectory:
+        segments.addAll(
+            fs.path.split(environment.cacheDir.resolveSymbolicLinksSync()));
+        break;
+      case Environment.kFlutterRootDirectory:
+        segments.addAll(
+            fs.path.split(environment.cacheDir.resolveSymbolicLinksSync()));
+        break;
+      default:
+        throw InvalidPatternException(pattern);
+    }
+    rawParts.skip(1).forEach(segments.add);
+    final String filePath = fs.path.joinAll(segments);
+    if (isDirectory) {
+      sources.add(SourceFile(fs.directory(fs.path.normalize(filePath))));
+    } else if (hasWildcard) {
+      // Perform a simple match by splitting the wildcard containing file on
```

Comment:
Note to have tests that use * in both appropriate and inappropriate ways.

Reply:
I've added some more test cases here. Longer term I'd like to find a way to use glob